packages/modular-scripts/src/build.ts

```diff
@@ -0,0 +1,488 @@
+  }
+}
+
+const publicPackageJsons: {
+  [name: string]: PackageJson;
+} = {};
+
+const typescriptConfig: TSConfig = {};
+// validate tsconfig
+{
+  // Extract configuration from config file and parse JSON,
+  // after removing comments. Just a fancier JSON.parse
+  const result = ts.parseConfigFileTextToJson(
+    typescriptConfigFilename,
+    fse.readFileSync(typescriptConfigFilename, 'utf8').toString(),
+  );
+
+  const configObject = result.config as TSConfig;
+
+  if (!configObject) {
+    reportTSDiagnostics([result.error as ts.Diagnostic]);
+    throw new Error('Failed to load Typescript configuration');
+  }
+
+  Object.assign(typescriptConfig, configObject);
+
+  Object.assign(typescriptConfig.compilerOptions, {
+    declarationDir: outputDirectory,
+    noEmit: false,
+    declaration: true,
+    emitDeclarationOnly: true,
+  });
+
+  Object.assign(typescriptConfig, {
+    exclude: [
+      // all TS test files, regardless whether co-located or in test/ etc
+      '**/*.stories.ts',
+      '**/*.stories.tsx',
+      '**/*.spec.ts',
+      '**/*.test.ts',
+      '**/*.e2e.ts',
+      '**/*.spec.tsx',
+      '**/*.test.tsx',
+      '__tests__',
+      // TS defaults below
+      'node_modules',
+      'bower_components',
+      'jspm_packages',
+      'tmp',
+    ],
+  });
+}
+
+async function makeBundle(directoryName: string): Promise<boolean> {
+  // TODO: - run whatever's in its scripts.build field too?
+  // TODO: verify that directoryName is in packageDirectoryNames
+
+  const packageJson = packageJsonsByDirectoryName[directoryName];
+
+  if (!packageJson) {
+    console.log(`no package.json in packages/${directoryName}, bailing...`);
+    return false;
+  }
+  if (packageJson.private === true) {
+    console.log(`packages/${directoryName} is marked private, bailing...`);
+    return false;
+  }
+
+  console.log(
+    `building ${packageJson.name as string} at packages/${directoryName}...`,
+  );
+
+  const bundle = await rollup.rollup({
+    // TODO: verify that .main exists
+    input: path.join(packagesRoot, directoryName, packageJson.main as string),
+    external: (id) => {
+      // via tsdx
+      if (id === 'babel-plugin-transform-async-to-promises/helpers') {
+        // we want to inline these helpers
+        return false;
+      }
+      // exclude any dependency that's not a realtive import
+      return !id.startsWith('.') && !path.isAbsolute(id);
+    },
+    treeshake: {
+      // via tsdx: Don't use getters and setters on plain objects.
+      propertyReadSideEffects: false,
+    },
+    plugins: [
+      resolve({
+        extensions,
+        browser: true,
+        mainFields: ['module', 'main', 'browser'],
+      }),
+      commonjs({ include: /\/node_modules\// }),
+      babel({
+        babelHelpers: 'bundled',
+        presets: [
+          ['@babel/preset-typescript', { isTSX: true, allExtensions: true }],
+          '@babel/preset-react',
+          [
+            '@babel/preset-env',
```

Comment:
Another important consideration here is overbundling: https://web.dev/publish-modern-javascript/ and https://next.material-ui.com/guides/minimizing-bundle-size/#modern-bundle when apps inside the monorepo are built, are they going to use output produced by this build when they point at packages in the monorepo?

Reply:
No, they use the source directly.

Reply:
That's good. Although even for published packages we should perhaps set a lower bound and not transpile below IE11, for example.